modules/qoa/audio_stream_qoa.cpp

```diff
@@ -0,0 +1,283 @@
+			//End of file or loop
+			if (qoa_stream->loop_mode == AudioStreamQOA::LOOP_FORWARD) {
+				seek(double(begin_limit / qoa_stream->mix_rate));
+				loops++;
+			} else {
+				frames_mixed_this_step = p_frames - todo;
+				//fill remainder with silence
+				for (int i = p_frames - todo; i < p_frames; i++) {
+					p_buffer[i] = AudioFrame(0, 0);
+				}
+				active = false;
+				todo = 0;
+			}
+		}
+	}
+	return frames_mixed_this_step;
+}
+
+float AudioStreamPlaybackQOA::get_stream_sampling_rate() {
+	return qoa_stream->mix_rate;
+}
+
+void AudioStreamPlaybackQOA::start(double p_from_pos) {
+	active = true;
+	seek(p_from_pos);
+	loops = 0;
+	begin_resample();
+}
+
+void AudioStreamPlaybackQOA::stop() {
+	active = false;
+}
+
+bool AudioStreamPlaybackQOA::is_playing() const {
+	return active;
+}
+
+int AudioStreamPlaybackQOA::get_loop_count() const {
+	return loops;
+}
+
+double AudioStreamPlaybackQOA::get_playback_position() const {
+	return double(frames_mixed) / qoa_stream->mix_rate;
+}
+
+void AudioStreamPlaybackQOA::seek(double p_time) {
+	if (!active) {
+		return;
+	}
+
+	if (p_time >= qoa_stream->get_length()) {
+		p_time = 0;
+	}
+
+	frames_mixed = uint32_t(qoa_stream->mix_rate * p_time);
+	decoded_len = 0;
+	decoded_offset = 0;
+
+	data_offset = 8 + frames_mixed / QOA_FRAME_LEN * qoa_max_frame_size(qoad);
+}
+
+void AudioStreamPlaybackQOA::tag_used_streams() {
+	qoa_stream->tag_used(get_playback_position());
+}
+
+AudioStreamPlaybackQOA::~AudioStreamPlaybackQOA() {
+	if (qoad)
+		memfree(qoad);
+	if (frame_data)
+		memfree(frame_data);
+	if (decoded)
+		memfree(decoded);
+}
+
+Ref<AudioStreamPlayback> AudioStreamQOA::instantiate_playback() {
+	Ref<AudioStreamPlaybackQOA> qoas;
+
+	ERR_FAIL_COND_V_MSG(data.is_empty(), qoas,
+			"This AudioStreamQOA does not have an audio file assigned "
+			"to it. AudioStreamQOA should not be created from the "
+			"inspector or with `.new()`. Instead, load an audio file.");
+
+	qoas.instantiate();
+	qoas->qoa_stream = Ref<AudioStreamQOA>(this);
+
+	qoas->qoad = (qoa_desc *)memalloc(sizeof(qoa_desc));
+	qoa_decode_header(data.ptr(), QOA_MIN_FILESIZE, qoas->qoad);
+
+	qoas->frame_data_len = qoa_max_frame_size(qoas->qoad);
+
+	qoas->frame_data = (unsigned char *)memalloc(qoas->frame_data_len);
+	qoas->decoded = (short *)memalloc(qoas->qoad->channels * QOA_FRAME_LEN * sizeof(short) * 2);
+
+	qoas->data_offset = 0;
+	qoas->frames_mixed = 0;
+	qoas->active = false;
+	qoas->loops = 0;
+
+	if (!qoas->qoad) {
+		ERR_FAIL_COND_V(!qoas->qoad, Ref<AudioStreamPlaybackQOA>());
+	}
+
+	return qoas;
+}
+
+String AudioStreamQOA::get_stream_name() const {
+	return "";
+}
+
+void AudioStreamQOA::clear_data() {
+	data.clear();
+}
+
+void AudioStreamQOA::set_data(const Vector<uint8_t> &p_data) {
+	int src_data_len = p_data.size();
+	const uint8_t *src_datar = p_data.ptr();
+
+	qoa_desc qoad;
+	qoad.channels = 0;
+	qoad.samplerate = 0;
+	uint32_t ffp = qoa_decode_header(p_data.ptr(), QOA_MIN_FILESIZE, &qoad);
```

Comment:
```suggestion
	uint32_t ffp = qoa_decode_header(p_data.ptr(), size, &qoad);
```
Or it will crash trying to read empty buffers, you can see this in the test runners